File: sdk/javascript/src/facade/NemFacade.js
/* eslint-disable no-unused-vars */
import { Bip32Node } from '../Bip32.js';
/* eslint-enable no-unused-vars */
import {
	Hash256,
	PrivateKey,
	/* eslint-disable no-unused-vars */
	PublicKey,
	SharedKey256,
	Signature
	/* eslint-enable no-unused-vars */
} from '../CryptoTypes.js';
import { NetworkLocator } from '../Network.js';
import { KeyPair, Verifier } from '../nem/KeyPair.js';
import MessageEncoder from '../nem/MessageEncoder.js';
import {
	Address,
	Network,
	/* eslint-disable no-unused-vars */
	NetworkTimestamp
	/* eslint-enable no-unused-vars */
} from '../nem/Network.js';
import { deriveSharedKey } from '../nem/SharedKey.js';
import TransactionFactory from '../nem/TransactionFactory.js';
/* eslint-disable no-unused-vars */
import * as nc from '../nem/models.js';
/* eslint-enable no-unused-vars */
import { keccak_256 } from '@noble/hashes/sha3.js';

// region NemPublicAccount / NemAccount

/**
 * NEM public account.
 */
export class NemPublicAccount {
	/**
	 * Creates a NEM public account.
	 * @param {NemFacade} facade NEM facade.
	 * @param {PublicKey} publicKey Account public key.
	 */
	constructor(facade, publicKey) {
		/**
		 * @protected
		 */
		this._facade = facade;

		/**
		 * Account public key.
		 * @type {PublicKey}
		 */
		this.publicKey = publicKey;

		/**
		 * Account address.
		 * @type {Address}
		 */
		this.address = this._facade.network.publicKeyToAddress(this.publicKey);
	}
}

/**
 * NEM account.
 */
export class NemAccount extends NemPublicAccount {
	/**
	 * Creates a NEM account.
	 * @param {NemFacade} facade NEM facade.
	 * @param {KeyPair} keyPair Account key pair.
	 */
	constructor(facade, keyPair) {
		super(facade, keyPair.publicKey);

		/**
		 * Account key pair.
		 * @type {KeyPair}
		 */
		this.keyPair = keyPair;
	}

	/**
	 * Creates a message encoder that can be used for encrypting and encoding messages between two parties.
	 * @returns {MessageEncoder} Message encoder using this account as one party.
	 */
	messageEncoder() {
		return new MessageEncoder(this.keyPair);
	}

	/**
	 * Signs a NEM transaction.
	 * @param {nc.Transaction} transaction Transaction object.
	 * @returns {Signature} Transaction signature.
	 */
	signTransaction(transaction) {
		return this._facade.signTransaction(this.keyPair, transaction);
	}
}

// endregion

/**
 * Facade used to interact with NEM blockchain.
 */
export class NemFacade {
	/**
	 * BIP32 curve name.
	 * @type {string}
	 */
	static BIP32_CURVE_NAME = 'ed25519-keccak';

	/**
	 * Network address class type.
	 * @type {typeof Address}
	 */
	static Address = Address;

	/**
	 * Network key pair class type.
	 * @type {typeof KeyPair}
	 */
	static KeyPair = KeyPair;

	/**
	 * Network verifier class type.
	 * @type {typeof Verifier}
	 */
	static Verifier = Verifier;

	/**
	 * Derives shared key from key pair and other party's public key.
	 * @param {KeyPair} keyPair Key pair.
	 * @param {PublicKey} otherPublicKey Other party's public key.
	 * @returns {SharedKey256} Shared encryption key.
	 */
	static deriveSharedKey = deriveSharedKey;

	/**
	 * Creates a NEM facade.
	 * @param {string|Network} network NEM network or network name.
	 */
	constructor(network) {
		/**
		 * Underlying network.
		 * @type {Network}
		 */
		this.network = 'string' === typeof network ? NetworkLocator.findByName(Network.NETWORKS, network) : network;

		/**
		 * Underlying transaction factory.
		 * @type {TransactionFactory}
		 */
		this.transactionFactory = new TransactionFactory(this.network);
	}

	/**
	 * Gets class type.
	 * @returns {typeof NemFacade} Class type.
	 */
	get static() { // eslint-disable-line class-methods-use-this
		return NemFacade;
	}

	/**
	 * Creates a network timestamp representing the current time.
	 * @returns {NetworkTimestamp} Network timestamp representing the current time.
	 */
	now() {
		return this.network.fromDatetime(new Date());
	}

	/**
	 * Creates a NEM public account from a public key.
	 * @param {PublicKey} publicKey Account public key.
	 * @returns {NemPublicAccount} NEM public account.
	 */
	createPublicAccount(publicKey) {
		return new NemPublicAccount(this, publicKey);
	}

	/**
	 * Creates a NEM account from a private key.
	 * @param {PrivateKey} privateKey Account private key.
	 * @returns {NemAccount} NEM account.
	 */
	createAccount(privateKey) {
		return new NemAccount(this, new KeyPair(privateKey));
	}

	/**
	 * Creates a transaction from a (typed) transaction descriptor.
	 * @param {object} typedDescriptor Transaction (typed) descriptor.
	 * @param {PublicKey} signerPublicKey Signer public key.
	 * @param {bigint} fee Transaction fee.
	 * @param {number} deadlineSeconds Approximate seconds from now for deadline.
	 * @returns {nc.Transaction} Created transaction.
	 */
	createTransactionFromTypedDescriptor(typedDescriptor, signerPublicKey, fee, deadlineSeconds) {
		const now = this.now();
		const transaction = this.transactionFactory.create({
			...typedDescriptor.toMap(),

			signerPublicKey,
			fee,
			timestamp: now.timestamp,
			deadline: now.addSeconds(deadlineSeconds).timestamp
		});
		return transaction;
	}

	// the following three functions are NOT static in order for NemFacade and SymbolFacade to conform to the same interface

	/**
	 * Hashes a NEM transaction.
	 * @param {nc.Transaction} transaction Transaction object.
	 * @returns {Hash256} Transaction hash.
	 */
	hashTransaction(transaction) { // eslint-disable-line class-methods-use-this
		const nonVerifiableTransaction = TransactionFactory.toNonVerifiableTransaction(transaction);
		return new Hash256(keccak_256(nonVerifiableTransaction.serialize()));
	}

	/**
	 * Gets the payload to sign given a NEM transaction.
	 * @param {nc.Transaction} transaction Transaction object.
	 * @returns {Uint8Array} Verifiable data to sign.
	 */
	extractSigningPayload(transaction) { // eslint-disable-line class-methods-use-this
		const nonVerifiableTransaction = TransactionFactory.toNonVerifiableTransaction(transaction);
		return nonVerifiableTransaction.serialize();
	}

	/**
	 * Signs a NEM transaction.
	 * @param {KeyPair} keyPair Key pair.
	 * @param {nc.Transaction} transaction Transaction object.
	 * @returns {Signature} Transaction signature.
	 */
	signTransaction(keyPair, transaction) {
		return keyPair.sign(this.extractSigningPayload(transaction));
	}

	/**
	 * Verifies a NEM transaction.
	 * @param {nc.Transaction} transaction Transaction object.
	 * @param {Signature} signature Signature to verify.
	 * @returns {boolean} \c true if transaction signature is verified.
	 */
	verifyTransaction(transaction, signature) {
		return new Verifier(transaction.signerPublicKey).verify(this.extractSigningPayload(transaction), signature);
	}

	/**
	 * Creates a network compatible BIP32 path for the specified account.
	 * @param {number} accountId Id of the account for which to generate a BIP32 path.
	 * @returns {Array<number>} BIP32 path for the specified account.
	 */
	bip32Path(accountId) {
		return [44, 'mainnet' === this.network.name ? 43 : 1, accountId, 0, 0];
	}

	/**
	 * Derives a NEM KeyPair from a BIP32 node.
	 * @param {Bip32Node} bip32Node BIP32 node.
	 * @returns {KeyPair} Derived key pair.
	 */
	static bip32NodeToKeyPair(bip32Node) {
		// BIP32 private keys should be used as is, so reverse here to counteract reverse in KeyPair
		const reversedPrivateKeyBytes = new Uint8Array([...bip32Node.privateKey.bytes]);
		reversedPrivateKeyBytes.reverse();

		return new KeyPair(new PrivateKey(reversedPrivateKeyBytes));
	}
}

File: sdk/java/src/main/java/org/symbol/sdk/facade/NemFacade.java
package org.symbol.sdk.facade;

import java.time.Instant;
import java.util.HashMap;
import java.util.Map;

import org.symbol.sdk.Bip32;
import org.symbol.sdk.CryptoTypes;
import org.symbol.sdk.JsonDescriptor;
import org.symbol.sdk.Network.NetworkLocator;
import org.symbol.sdk.nem.Address;
import org.symbol.sdk.nem.KeyPair;
import org.symbol.sdk.nem.MessageEncoder;
import org.symbol.sdk.nem.NemTransactionFactory;
import org.symbol.sdk.nem.Network;
import org.symbol.sdk.nem.NetworkTimestamp;
import org.symbol.sdk.nem.Verifier;
import org.symbol.sdk.nem.descriptors.NemTransactionDescriptor;
import org.symbol.sdk.nem.models.*;
import org.symbol.sdk.utils.ArrayHelpers;
import org.symbol.sdk.utils.Transforms;

/**
 * High-level facade used to interact with the NEM blockchain: signing, verifying, hashing, and creating transactions.
 */
public final class NemFacade {

	/** BIP32 curve name for NEM. */
	public static final String BIP32_CURVE_NAME = "ed25519-keccak";

	/** Underlying network. */
	public final Network network;

	/** Underlying transaction factory. */
	public final NemTransactionFactory transactionFactory;

	/**
	 * Creates a NEM facade around an explicit network.
	 *
	 * @param network NEM network.
	 */
	public NemFacade(final Network network) {
		this.network = network;
		this.transactionFactory = new NemTransactionFactory(network);
	}

	/**
	 * Creates a NEM facade by network name.
	 *
	 * @param networkName Network name (e.g. {@code "mainnet"}, {@code "testnet"}).
	 */
	public NemFacade(final String networkName) {
		this(NetworkLocator.findByName(Network.NETWORKS, networkName));
	}

	/**
	 * Creates a network timestamp representing the current time.
	 *
	 * @return Network timestamp representing the current time.
	 */
	public NetworkTimestamp now() {
		return network.fromDatetime(Instant.now());
	}

	/**
	 * Creates a NEM public account from a public key.
	 *
	 * @param publicKey Account public key.
	 * @return NEM public account.
	 */
	public NemPublicAccount createPublicAccount(final CryptoTypes.PublicKey publicKey) {
		return new NemPublicAccount(this, publicKey);
	}

	/**
	 * Creates a NEM account from a private key.
	 *
	 * @param privateKey Account private key.
	 * @return NEM account.
	 */
	public NemAccount createAccount(final CryptoTypes.PrivateKey privateKey) {
		return new NemAccount(this, new KeyPair(privateKey));
	}

	/**
	 * Creates a transaction from a descriptor, adding signer, fee, timestamp and deadline to a copy of it.
	 *
	 * @param descriptor Transaction descriptor (must contain {@code "type"}); not mutated.
	 * @param signerPublicKey Signer public key.
	 * @param fee Transaction fee.
	 * @param deadlineSeconds Approximate seconds from now for the deadline.
	 * @return Created transaction.
	 */
	Transaction createTransactionFromDescriptor(final Map<String, Object> descriptor, final CryptoTypes.PublicKey signerPublicKey,
			final long fee, final long deadlineSeconds) {
		final NetworkTimestamp current = now();
		final Map<String, Object> rawDescriptor = new HashMap<>(descriptor);
		rawDescriptor.put("signerPublicKey", signerPublicKey);
		rawDescriptor.put("fee", fee);
		rawDescriptor.put("timestamp", current.timestamp);
		rawDescriptor.put("deadline", current.addSeconds(deadlineSeconds).timestamp);
		return transactionFactory.create(rawDescriptor);
	}

	/**
	 * Creates a transaction from a typed transaction descriptor.
	 *
	 * @param typedDescriptor Typed transaction descriptor.
	 * @param signerPublicKey Signer public key.
	 * @param fee Transaction fee.
	 * @param deadlineSeconds Approximate seconds from now for the deadline.
	 * @return Created transaction.
	 */
	public Transaction createTransactionFromTypedDescriptor(final NemTransactionDescriptor typedDescriptor,
			final CryptoTypes.PublicKey signerPublicKey, final long fee, final long deadlineSeconds) {
		return createTransactionFromDescriptor(typedDescriptor.toMap(), signerPublicKey, fee, deadlineSeconds);
	}

	/**
	 * Creates a transaction from a JSON descriptor document.
	 *
	 * @param json JSON object with the same shape as the untyped descriptor map.
	 * @param signerPublicKey Signer public key.
	 * @param fee Transaction fee.
	 * @param deadlineSeconds Approximate seconds from now for the deadline.
	 * @return Created transaction.
	 */
	public Transaction createTransactionFromJson(final String json, final CryptoTypes.PublicKey signerPublicKey, final long fee,
			final long deadlineSeconds) {
		return createTransactionFromDescriptor(JsonDescriptor.parse(json), signerPublicKey, fee, deadlineSeconds);
	}

	// the following three functions are NOT static in order for NemFacade and SymbolFacade to conform to the same interface

	/**
	 * Hashes a NEM transaction; the hash is Keccak-256 of the non-verifiable serialization.
	 *
	 * @param transaction Transaction object.
	 * @return Transaction hash.
	 */
	public CryptoTypes.Hash256 hashTransaction(final Transaction transaction) {
		final byte[] payload = NemTransactionFactory.toNonVerifiableTransaction(transaction).serialize();
		return new CryptoTypes.Hash256(Transforms.keccak_256(payload));
	}

	/**
	 * Gets the payload to sign for a NEM transaction.
	 *
	 * @param transaction Transaction object.
	 * @return Verifiable data to sign — the non-verifiable serialization of the transaction.
	 */
	public byte[] extractSigningPayload(final Transaction transaction) {
		return NemTransactionFactory.toNonVerifiableTransaction(transaction).serialize();
	}

	/**
	 * Signs a NEM transaction.
	 *
	 * @param keyPair Signer key pair.
	 * @param transaction Transaction object.
	 * @return Transaction signature.
	 */
	public CryptoTypes.Signature signTransaction(final KeyPair keyPair, final Transaction transaction) {
		return keyPair.sign(extractSigningPayload(transaction));
	}

	/**
	 * Verifies a NEM transaction signature.
	 *
	 * @param transaction Transaction object.
	 * @param signature Signature to verify.
	 * @return {@code true} if the signature is valid for the transaction.
	 */
	public boolean verifyTransaction(final Transaction transaction, final CryptoTypes.Signature signature) {
		return new Verifier(new CryptoTypes.PublicKey(transaction.getSignerPublicKey().bytes())).verify(extractSigningPayload(transaction),
				signature);
	}

	/**
	 * Creates a network-compatible BIP32 path for the specified account.
	 *
	 * @param accountId Id of the account for which to generate a BIP32 path.
	 * @return BIP32 path for the specified account.
	 */
	public int[] bip32Path(final int accountId) {
		final int coinType = "mainnet".equals(network.name) ? 43 : 1;
		return new int[]{
				44, coinType, accountId, 0, 0
		};
	}

	/**
	 * Derives a NEM {@link KeyPair} from a BIP32 node. BIP32 private keys should be used as-is, so the bytes are reversed here to
	 * counteract the NEM {@link KeyPair} constructor's internal reverse.
	 *
	 * @param bip32Node BIP32 node.
	 * @return Derived key pair.
	 */
	public static KeyPair bip32NodeToKeyPair(final Bip32.Bip32Node bip32Node) {
		return new KeyPair(new CryptoTypes.PrivateKey(ArrayHelpers.reverse(bip32Node.privateKey.bytes())));
	}

	// region account wrappers

	/**
	 * NEM public account — known public key plus derived address.
	 */
	public static class NemPublicAccount implements PublicAccount {
		/** Owning facade. */
		protected final NemFacade facade;

		/** Account public key. */
		private final CryptoTypes.PublicKey publicKey;

		/** Account address. */
		private final Address address;

		NemPublicAccount(final NemFacade facade, final CryptoTypes.PublicKey publicKey) {
			this.facade = facade;
			this.publicKey = publicKey;
			this.address = facade.network.publicKeyToAddress(publicKey);
		}

		@Override
		public CryptoTypes.PublicKey publicKey() {
			return publicKey;
		}

		@Override
		public Address address() {
			return address;
		}
	}

	/**
	 * NEM account — adds the signing key pair to {@link NemPublicAccount}.
	 */
	public static final class NemAccount extends NemPublicAccount implements Account<Transaction, KeyPair> {
		/** Account key pair. */
		private final KeyPair keyPair;

		NemAccount(final NemFacade facade, final KeyPair keyPair) {
			super(facade, keyPair.getPublicKey());
			this.keyPair = keyPair;
		}

		@Override
		public KeyPair keyPair() {
			return keyPair;
		}

		/**
		 * Creates a {@link MessageEncoder} using this account as one party.
		 *
		 * @return Message encoder.
		 */
		public MessageEncoder messageEncoder() {
			return new MessageEncoder(this.keyPair);
		}

		/**
		 * Signs a NEM transaction.
		 *
		 * @param transaction Transaction object.
		 * @return Transaction signature.
		 */
		public CryptoTypes.Signature signTransaction(final Transaction transaction) {
			return this.facade.signTransaction(this.keyPair, transaction);
		}
	}

	// endregion
}
